src/main.rs

```diff
@@ -1,126 +1,267 @@
-use nalgebra::Point2D; //TODO: Add the nalgebra crate and lookup the closest struct to Point2D
+use std::f32::consts::PI;
+use std::time::Duration;
+use std::path::Path;
+use std::thread;
+use std::rc::Rc;
 
-// TODO: Pick a better name than AppError
-#[derive(Debug)]
-enum AppError {
-    IOError(IOError),
-    FBXError(FBXError),
+use euc::{Pipeline, rasterizer, buffer::Buffer2d, Target};
+use minifb::{self, Key, KeyRepeat};
+use tobj;
+use vek::{Mat4, Vec2, Vec3, Vec4, Rgba};
+use image::ImageBuffer;
+
+use crate::cel::CelShader;
+use crate::outline::OutlineShader;
+use crate::light::DiffuseLight;
+use crate::scale::scale_buffer;
+use crate::geometry::Mesh;
+use crate::material::Material;
+use crate::color::{rgba_to_bgra_u32, bgra_u32_to_rgba, vek_rgba_to_image_rgba};
+
+fn main() {
+    let image_width = 64;
+    let image_height = 64;
+
+    let frames: Vec<_> = (1..=8).map(|i| {
+        load_frame(&format!("samples/bigboi/obj/bigboi_rigged_{:06}.obj", i))
+    }).collect();
+
+    // The transformation that represents the center of the model, all points in the model are
+    // relative to this
+    // Also known as the "world" transformation
+    //
+    // Model coordinates -> World coordinates
+    let model = Mat4::identity();
+    // The transformation that represents the position and orientation of the camera
+    //
+    // World coordinates -> Camera coordinates
+    let view = Mat4::rotation_x(PI/8.0) * Mat4::rotation_y(0.0*PI/2.0);
+    // The perspective/orthographic/etc. projection of the camera
+    //
+    // Camera coordinates -> Homogenous coordinates
+    let projection = Mat4::perspective_rh_no(0.8*PI, (image_width as f32)/(image_height as f32), 0.01, 100.0)
+        * Mat4::<f32>::scaling_3d(0.6);
+
+    save_poses(image_width, image_height, model, view, projection, &frames);
+    save_spritesheet(image_width, image_height, model, view, projection, &frames);
+    preview_window(image_width, image_height, model, view, projection, &frames);
 }
 
-impl From<IOError> for AppError {
-    fn from(error: IOError) -> Self {
-        AppError::IOError(error)
+fn save_poses(
+    image_width: usize,
+    image_height: usize,
+    model: Mat4<f32>,
+    view: Mat4<f32>,
+    projection: Mat4<f32>,
+    frames: &[Vec<Mesh>],
+) {
+    let mut color = Buffer2d::new([image_width, image_height], 0);
+    let mut depth = Buffer2d::new([image_width, image_height], 1.0);
+
+    for (i, frame) in frames.into_iter().enumerate() {
+        render(&mut color, &mut depth, model, view, projection, frame, 0.15);
+
+        let mut img = ImageBuffer::new(image_width as u32, image_height as u32);
+        for (x, y, pixel) in img.enumerate_pixels_mut() {
+            // Unsafe because we are guaranteeing that these indexes are not out of bounds
+            let color = unsafe { color.get([x as usize, y as usize]) };
+            let rgba = bgra_u32_to_rgba(*color);
+            *pixel = vek_rgba_to_image_rgba(rgba);
```

Comment:
why did we need to store pixel colors as BGRA elements in the first place? It seems like we process RGBA colors in the pixel shader, convert to BGRA, and now we're converting it back to RGBA it seems 😮

Reply:
Agreed! I will likely change this at a later date (maybe in a separate PR) to just keep everything in RGBA. The reason we do this (that I also mentioned on the call) is because of the windowing library we use. minifb exposes a frame buffer that expects us to use BGRA colors represented as 32-bit unsigned integers. Given that we don't want to couple so tightly to the GUI library, it's probably best to stop doing this in our shader.